From 0bbe3a9643219b40f37cae0a41733bf9b48325a6 Mon Sep 17 00:00:00 2001
From: Andreas Rammhold <andreas@rammhold.de>
Date: Mon, 1 Jun 2020 22:09:53 +0200
Subject: [PATCH 7/7] Add some dhcpv6 debug statements

---
 src/network/networkd-dhcp6.c | 19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git a/src/network/networkd-dhcp6.c b/src/network/networkd-dhcp6.c
index cd6ac54224..235689e30f 100644
--- a/src/network/networkd-dhcp6.c
+++ b/src/network/networkd-dhcp6.c
@@ -295,8 +295,9 @@ static int dhcp6_pd_prefix_distribute(Link *dhcp6_link,
         n_prefixes = UINT64_C(1) << (64 - pd_prefix_len);
 
         (void) in_addr_to_string(AF_INET6, &prefix, &buf);
-        log_link_debug(dhcp6_link, "Assigning up to %" PRIu64 " prefixes from %s/%u",
-                       n_prefixes, strnull(buf), pd_prefix_len);
+        log_link_debug(dhcp6_link, "Assigning up to %" PRIu64 " prefixes from %s/%u for %s",
+                       n_prefixes, strnull(buf), pd_prefix_len,
+                       (assign_preferred_subnet_id ? "preferred subnet id's" : "'auto' subnet id"));
 
         HASHMAP_FOREACH(link, manager->links, i) {
                 union in_addr_union assigned_prefix;
@@ -307,15 +308,25 @@ static int dhcp6_pd_prefix_distribute(Link *dhcp6_link,
                 if (!dhcp6_get_prefix_delegation(link))
                         continue;
 
-                if (dhcp6_link_has_dhcpv6_prefix(link))
+                log_link_debug(link, "Searching for a suitable prefix");
+
+                if (dhcp6_link_has_dhcpv6_prefix(link)) {
+                        log_link_debug(link, "Already has a prefix. Skipping.");
                         continue;
+                }
+
+
+                if (assign_preferred_subnet_id != dhcp6_has_preferred_subnet_id(link)) {
+                        log_link_debug(link, "Skipping prefix search as this isn't the mode we are looking for");
 
-                if (assign_preferred_subnet_id != dhcp6_has_preferred_subnet_id(link))
                         continue;
+                }
 
                 r = dhcp6_get_preferred_delegated_prefix(manager, link, &prefix.in6, pd_prefix_len,
                                                         &assigned_prefix.in6);
 
+                log_link_debug(link, "dhcp6_get_preferred_delegated_prefix returned %i", r);
+
                 if (assign_preferred_subnet_id && r == -EAGAIN) {
                         /* A link has a preferred subnet_id but that one is
                          * already taken by another link. Now all the remaining
-- 
2.26.2

