From 61d08782fcdf3bc34a45fd046672da7f6c358ee9 Mon Sep 17 00:00:00 2001
From: Andreas Rammhold <andreas@rammhold.de>
Date: Tue, 2 Jun 2020 11:32:52 +0200
Subject: [PATCH 8/8] libsystemd-network: add logging to sd_radv_remove_prefix

---
 src/libsystemd-network/sd-radv.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/libsystemd-network/sd-radv.c b/src/libsystemd-network/sd-radv.c
index 873a2f40f8..6a974cede4 100644
--- a/src/libsystemd-network/sd-radv.c
+++ b/src/libsystemd-network/sd-radv.c
@@ -605,6 +605,7 @@ _public_ sd_radv_prefix *sd_radv_remove_prefix(sd_radv *ra,
                                                const struct in6_addr *prefix,
                                                unsigned char prefixlen) {
         sd_radv_prefix *cur, *next;
+        _cleanup_free_ char * addr_p = NULL;
 
         assert_return(ra, NULL);
         assert_return(prefix, NULL);
@@ -618,6 +619,9 @@ _public_ sd_radv_prefix *sd_radv_remove_prefix(sd_radv *ra,
                                    (union in_addr_union *)&cur->opt.in6_addr))
                         continue;
 
+                (void) in_addr_to_string(AF_INET6, (union in_addr_union*) prefix, &addr_p);
+                log_radv("Removed prefix %s/%d", strnull(addr_p), prefixlen);
+
                 LIST_REMOVE(prefix, ra->prefixes, cur);
                 ra->n_prefixes--;
                 sd_radv_prefix_unref(cur);
-- 
2.26.2

